docs/2024/puzzles/day21.md

```diff
@@ -2,10 +2,243 @@ import Solver from "../../../../../website/src/components/Solver.js"
+  val reverse = !positions(from + shift.projX) || (positions(from + shift.projY) && shift.x > 0)
+  if reverse then v + h + 'A' else h + v + 'A'
+
+def minPath(input: String, isNumerical: Boolean = false): String =
+  val keypad = if isNumerical then numericalKeypad else directionalKeypad
+  val positions = if isNumerical then numericalKeypadPositions else directionalKeypadPositions
+  (s"A$input").map(keypad).sliding(2).map(p => minPathStep(p(0), p(1), positions)).mkString
+
+def part1(input: String): Long =
+  input
+    .linesIterator
+    .filter(_.nonEmpty)
+    .map: line => // 029A
+      val path1 = minPath(line, isNumerical = true) // <A^A^^>AvvvA
+      val path2 = minPath(path1) // v<<A>>^A<A>A<AAv>A^A<vAAA^>A
+      val path3 = minPath(path2) // <vA<AA>>^AvAA<^A>Av<<A>>^AvA^Av<<A>>^AA<vA>A^A<A>Av<<A>A^>AAA<Av>A^A
+      val num = line.init.toLong // 29
+      val len = path3.length() // 68
+      len * num // 211930
+    .sum
+```
+
+The comments in `part1` demonstrate intermediate results for the sample input `029A`.
+
+## Part 2
+
+Although the above approach works for three consecutive robots, it does not scale to 25 robots because path size grows exponentially. Instead, we refactor the code to compute only the cost of each path, not the path itself. This leads to two new functions, `minPathStepCost` and `minPathCost`, which incorporate a `level` parameter for the current robot (with 0 indicating the numerical keypad) and a `maxLevel` parameter for the last robot. We also use [memoization](https://en.wikipedia.org/wiki/Memoization) to cache results for performance, since these functions are called repeatedly with the same arguments.
+
+```scala mdoc:silent
+val cache = collection.mutable.Map.empty[(Pos, Pos, Int, Int), Long]
+def minPathStepCost(from: Pos, to: Pos, level: Int, maxLevel: Int): Long =
+  cache.getOrElseUpdate((from, to, level, maxLevel), {
+    val positions = if level == 0 then numericalKeypadPositions else directionalKeypadPositions
+    val shift = to - from
+    val h = (if shift.x > 0 then ">" else "<") * shift.x.abs
+    val v = (if shift.y > 0 then "v" else "^") * shift.y.abs
+    val reverse = !positions(from + shift.projX) || (positions(from + shift.projY) && shift.x > 0)
+    val res = if reverse then v + h + 'A' else h + v + 'A'
+    if level == maxLevel then res.length() else minPathCost(res, level + 1, maxLevel)
+  })
+
+def minPathCost(input: String, level: Int, maxLevel: Int): Long =
+  val keypad = if level == 0 then numericalKeypad else directionalKeypad
+  (s"A$input").map(keypad).sliding(2).map(p => minPathStepCost(p(0), p(1), level, maxLevel)).sum
+
+def part2(input: String): Long =
+  input
+    .linesIterator
+    .filter(_.nonEmpty)
+    .map(line => minPathCost(line, 0, 25) * line.init.toLong)
+    .sum
+```
+
+On my example puzzle input, this solution runs in about 55 ms on the JVM and 3 ms on [Scala Native](https://scala-native.org/en/stable/).  
```

Comment:
Is most of that JVM startup time and/or HotSpot warmup time? If so, I'd suggest saying so, otherwise readers might get a mistaken impression.

And if not, I'm curious why Native would be so much faster.

In some contexts I would insist that any benchmark data be JMH-based (which is pretty easy with scala-cli's JMH support, actually) rather then going the `currentTimeMillis` route, but I don't want to make more work for you. But if it stays as is, I think it could use some additional wording about how seriously the numbers are to be taken.

Reply:
Performance section added ✅

Indeed, the main difference is due to the lack of warm-up. However, I believe cold runtimes are just as important as hot runtimes, since not every application can afford 50k iterations before real work begins. That’s especially relevant here, where the script in a one-off. I do agree, however, that reporting both is better for completeness.

Reply:
Wait, my JMH benchmarks make no sense, I forgot to reset the cache between iterations 😭 I'll fix that tomorrow.

Reply:
nerdsnipe succeeded 😈

nice improvement!